File: storage.cpp
#include <iostream>
#include <cstring>

using namespace std;

// Represent a stack list in C++
class Node
{
public:
    int data;
    Node *next;

    // Default constructor
    Node()
    {
        data = 0;
        next = NULL;
    }

    // Constructor with data
    Node(int data)
    {
        this->data = data;
        this->next = NULL;
    }

};

class Storage
{
public:
    Node *head;

    /** Default constructor
     * 
    */
    Storage()
    {
        head = NULL;
    }

    /** Create a Node to hold the data, then put it at the head of the list.
     * @param data The data for the new node
    */
    void push(int data);

    /** Remove the head Node and return its data.
     * 
     * @param [out] The data (only valid if return true)
     * @return True if there is a node to return
    */
    bool pop(int &data);

    /**
     * Return the data from the head Node, without removing it.
     * 
     * @param [out] The data (only valid if return true)
     * @return True if text has been returned successfully (there is a node)
    */    
    bool peek(int &data);

    /**
     * Return True if the list is empty, otherwise False  
     * 
     * @return True if the list is empty, otherwise False
    */
    bool isEmpty();

    /**
     * Swaps the nodes at position i and j.
     * 
     * @return Returns true if successful, otherwise false (e.g. outside range)
    */
    bool swap(int i, int j);
    

   /**
    * @brief Prints a given stack
    * 
    */
   void printList();

};

// Function that will add a new node to the storage
void Storage::push(int data)
{
    // allocate space for new node in heap
    Node *newNode = new Node(data);

    // check if stack/heap is full.
    // Inserting an element would
    // lead to stack overflow
    if(!newNode)
    {
        cerr << "Log of negative number is: " << errno << "\n";
        cerr << "Cause of error: " << strerror(errno) << endl;
        exit(-1);
    }

    // checks if storage is empty
    // and sets the new value to given push() parameter
    if(head == NULL)
    {
        head = newNode;
    }
    else
    {
        // next new node will be assigned as head like stack
        // last-in first-out (LIFO)
        // that head will point to new node
        newNode->next = head;
        head = newNode;
    }
}

// Function to remove a key from the beginning
bool Storage::pop(int &data)
{
    Node *newNode = head;

    if(newNode == NULL)
    {
        cout << "Stack is empty." << endl;
        return false;
    }
    else
    {
        cout << "The popped element is " << newNode->data << endl;
        head = newNode->next;
        return true;
    }
}

// Function that returns the top element in a stack
bool Storage::peek(int &data)
{
    Node *newNode = head;
    // if stack is empty
    if(head == NULL)
    {
        return false;
    }
    else
    {
        // if stack is not empty, return the top element
        cout << "Top element is " << newNode->data << endl;
        return true;
    }
}

// Function to check if the storage is empty or not
bool Storage::isEmpty()
{
    // If head is NULL it means that
    // there are no elements in the storage
    if(head == NULL)
    {
        cout << "Stack is empty" << endl;
        return true;
    }
    else
    {
        // if stack is filled
        cout << "Stack is not empty" << endl;
        return false;
    }

}

// Function to swap two node values
bool Storage::swap(int i, int j)
{
    // if i and j are equal, then the list will remain the same
    if(i == j)
    {
        return true;
    }

    // count how many nodes there are in the list
    int count = 0;
    Node *temp = head;
    while(temp != NULL)
    {
        count++;
        temp = temp->next;
    }

    Node *p1 = head;
    Node *p2 = head;

    // checks if storage is empty
    if(p1 && p2 == NULL)
    {
        cerr << "Swapping is not possible" << endl;
        return false;
    }

    // traverse to the nodes where given values to be swapped
    for(int k = 0; k < i; k++)
    {
        p1 = p1->next;
    }
    for(int l = 0; l < j; l++)
    {
        p2 = p2->next;
    }

    // swap the values of two nodes
    int val = p1->data;
    p1->data = p2->data;
    p2->data = val;

    return true;

}
// Function to print the nodes in linked list
void Storage::printList()
{
    Node *temp = head;

    if(head == NULL)
    {
        cerr << "Stack is empty" << endl;
    }
    else
    {
        // Traverse the list
        cout << "Stack elements are: ";
        cout << "None -> ";
        while(temp != NULL)
        {
            cout << temp->data << " -> ";
            temp = temp->next;
        }
    }
    cout << endl;

}

// Main function for execution
int main()
{
    // create an instance (object) of a class
    Storage *s = new Storage();
    int pop = 0, peek = 0;

    // appends values to the stack
    s->push(1);
    // prints the stack
    s->printList();
    s->push(2);
    s->printList();
    s->push(3);
    s->printList();
    s->push(4);
    s->printList();
    s->push(5);
    s->printList();
    // checks if the storage is empty or not
    s->isEmpty();
    // removes the node from the top
    s->pop(pop);
    s->printList();
    s->pop(pop);
    s->printList();
    s->push(10);
    s->printList();
    // prints the top node in the stack
    s->peek(peek);
    s->push(19);
    s->printList();
    s->peek(peek);
    s->pop(pop);
    s->printList();
    cout << "\nBefore swapping: " << "\n";
    s->printList();
    // swapping nodes at index 0 & 2
    s->swap(0, 2);
    cout << "\nAfter swapping: " << endl;
    s->printList();

    return 0;
}

File: storage.py
import sys
# Represent a stack list in Python3
class Node:
    # create nodes of stack
    # constructor using data
    def __init__(self, data):
        self.data = data
        self.next = None

    # returns a string representation of the data
    def __str__(self):
        return str(self.data)


class Storage:

    # default constructor
    # head is default NULL
    def __init__(self):
        self.head = None

    # adds data to the stack
    # adds data to the end of the stack
    def push(self, data):
        new_node = Node(data)

        if(not new_node):
            print("Stack overflow!!!")
            sys.exit(0);
        
        if self.head is None:
            self.head = new_node
        else:
            new_node.next = self.head
            self.head = new_node

    # Function to remove the first element/current head (start of the stack)
    def pop(self):
        if(self.head == None):
            return 'Stack is empty'
        else:
            # Removes the current head node
            # makes the new value the new head
            result = self.head.data
            self.head = self.head.next
            print(f"Popped value is: {result}")
            return result
    
    # Function that returns the top of element in a stack
    def peek(self):
        if (self.head == None):
            raise Exception("Peeking from an empty stack")
            return None
        else:
            new_node = self.head
            print(f"Top element is {new_node.data}")
            return new_node.data

    # Checks if stack is empty
    def isempty(self):
        if self.head is None:
            print("Stack is empty")
            return True
        else:
            print("Stack is not empty")
            return False
    
    # Function to swap the given two nodes
    def swap(self, i, j):
        # if i and j are equal or null, the list will remain the same
        if(i == j):
            return
        
        # count the number of nodes in the list
        count = 0
        temp = self.head
        while(temp is not None):
            count += 1
            temp = temp.next
        
        # check if the swap positions are valid
        if(i < 0 or i > count or j < 0 or j > count):
            return i, j
        
        p1 = self.head
        p2 = self.head

        # checks if storage is empty
        if(p1 and p2 is None):
            print("Swapping is not possible")

        # traverse to the nodes where given values to be swapped
        for i in range(1, i):
            p1 = p1.next
        for i in range(1, j):
            p2 = p2.next
        
        # swap the values of two nodes
        val = p1.data
        p1.data = p2.data
        p2.data = val

    # prints out the stack
    def __str__(self):
        node = self.head
        my_str = ""
        my_str += "None -> "

        while node is not None:
            my_str += f"{node.data} -> "
            node = node.next

        return my_str

# Driver code
if __name__ == '__main__':
    s = Storage()

    # Add new elements in the stack
    s.push(100)
    print(s)
    s.push(3)
    print(s)
    s.push(19)
    print(s)
    s.push(15)
    print(s)
    s.push(10)
    print(s)
    s.push(95)

    print(s)
    s.isempty()
    s.swap(1, 6)
    print(s)
    s.peek()
    s.pop()
    print(s)